### ISSUE TITLE
Add TypeScript typings for 'meteor/apollo'

### ISSUE THREAD

**Issue by KeithGillette** [apollographql/meteor-integration#65]
Are there any TypeScript type definitions available for `meteor-integration` to overcome the TypeScript `TS2307:Cannot find module 'meteor/apollo'.` warning message? Any work-arounds? Thanks!

**Comment by xavcz** [comment]
Hey @KeithGillette, did you find a workaround in the end?

How are you using TypeScript in your Meteor application?

**Comment by KeithGillette** [comment]
> Hey @KeithGillette, did you find a workaround in the end?

Hi @xavcz: I have not found a work-around and still get the TypeScript warning.

> How are you using TypeScript in your Meteor application?

I'm not sure I understand the question. Do you mean "What's your TypeScript-meteor build tooling?" or "How are you leveraging TypeScript to build a Meteor application?"

**Comment by xavcz** [comment]
So #84 won't help.

@lorensr much better understanding of TS now, cheers 🙆‍♂️ 

@KeithGillette indeed, the types for `meteor/apollo` should be accessible from NPM. 

Do you think you can make a PR to https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/meteor  to create the 'meteor/apollo' module and the related functions (`createMeteorNetworkInterface`, `meteorClientConfig`, `createApolloServer`)? Or maybe to a new folder there?

**Comment by KeithGillette** [comment]
Thanks for the follow-up @xavcz. I'd love to help get a fix for this, but am not sure how to proceed.

According to the [README for the Definitely Typed Meteor Type Definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/meteor), they are automatically created  from compiled Meteor code, so I don't think it would be welcome to just insert typings for `meteor/apollo` into the Definitely Typed meteor folder via a PR. However,ccording to the [meteor-typings/meteor README], the `meteor-typings` organization is supposed to contain all available typings for Atmosphere packages, so perhaps that is the place to submit a pull request.

Further, I have not been able to get the `TS2307:Cannot find module 'meteor/apollo'` to go away entirely. I was able to prevent it from being output during a `meteor run` by creating a `meteor.d.ts` file with the following contents, but I still get the error in files that contain the `import from `meteor/apollo` and TypeScript does not provide intellisense for the definitions you provided.

```TypeScript
import { NetworkInterface } from 'apollo-client';

declare module apollo {

	function createMeteorNetworkInterface(
		customNetworkInterfaceConfig: any
	): NetworkInterface;

	function meteorClientConfig(
		customClientConfig: any
	): any;

	function createApolloServer(
		customOptions: any,
		customConfig: any
	): void;
}
```

I am not conversant enough with creating TypeScript declaration files to know what's wrong.

**Comment by csillag** [comment]
Three things:

1. `ApolloClientConfig` should be start with an upper case

2. I think most of the fields here should be optional:

``` TypeScript
export interface customApolloServerConfig {
	path?: string;
	configServer?: () => any;
	graphiql?: boolean;
	graphiqlPath?: string;
	graphiqlOptions?: graphiqlOptions;
}
```

3. Furthermore, I would add the following:

``` TypeScript
export interface ApolloServerOptions {
    schema: GraphQLSchema,
    context?: any,
    rootValue?: any,
    formatError?: Function,
    validationRules?: Array<ValidationRule>,
    formatParams?: Function,
    formatResponse?: Function,
    debug?: boolean,
}
 
export const createApolloServer: (options: ApolloServerOptions, config?: ApolloServerConfig) => void;
export const meteorClientConfig: (): ApolloClientConfig;        
```

**Comment by KeithGillette** [comment]
Thanks, @csillag. You're probably right about points 1 & 2. With regard to 3, I was just following the naming convention I found in the documentation for [Apollo > Server > GraphQL server options](http://dev.apollodata.com/tools/graphql-server/setup.html#graphqlOptions), referenced in the [Apollo > Meteor > API > create ApolloServer](http://dev.apollodata.com/core/meteor.html#createApolloServer) documentation, which uses `GraphQLOptions` instead of `ApolloServerOptions` (the latter being more parallel to `ApolloServerConfig` and my preference, as well). However, I'd defer it to the maintainers of this repository to choose final naming.

Regardless, the question still remains how best to get these typings packaged and published. As I said before, it looks to me like the [`meteor-typings`](https://github.com/meteor-typings) organization might be the best spot to house them, but @barbatus hasn't weighed in on that. I see that @exKAZUu is the only person associated with that organization, so I'm tagging him in hopes he can provide some guidance.

**Comment by KeithGillette** [comment]
I tweaked the ambient type definitions file and created a [Apollo-GraphQL-Meteor-Integration-Typings](https://github.com/KeithGillette/Apollo-GraphQL-Meteor-Integration-Typings) GitHub repository to share them. I started to follow the [DefinitelyType contribution guide](http://definitelytyped.org/guides/contributing.html) in hopes these can at some point be included in the DefinitelyTyped repository, but did not get to creating any tests per their requirements.

Regardless, it still seems to me that these should be housed in the [meteor-typings](https://github.com/meteor-typings) organization, but I have not heard from @barbatus or @exKAZUu for guidance on that question.

**Comment by exKAZUu** [comment]
@KeithGillette I'm not sure what is the best way. Anyway, I can create a repo in meteor-typings organization for you if you want me to do so.

**Comment by KeithGillette** [comment]
If you don't know, @exKAZUu, then who does?! ;-)

I was just going by the [Typings for Meteor README](https://github.com/meteor-typings/meteor#atmosphere-packages), which states:

> If you want to install some Atmosphere package's typings, search for an appropriate repo in the current meteor-typings organization, which is supposed to contain all available typings of such type.
If you have found that package's typings repo, go inside and read how to install typings. If not - you are very welcome to create new repo and contribute typings.

Are the repositories in `meteor-typings` only for ambient TypeScript definitions distributed via [Typings](https://github.com/typings/registry) or installed manually? Is there any automatic build that you do to get the typings in `meteor-typings` into [DefinitelyTyped](http://definitelytyped.org/)?

**Comment by xavcz** [comment]
Looks like there is no progress on the `DefinitelyTyped` thing? Let's add a link to (your repo)[https://github.com/KeithGillette/Apollo-GraphQL-Meteor-Integration-Typings] in the docs 👌

**Comment by KeithGillette** [comment]
Thanks for the follow-up @xavcz.

I haven't gotten any substantive guidance from the maintainers of the [`meteor-typings`](https://github.com/meteor-typings) organization on how they manage submission to [`DefinitelyTyped`](http://definitelytyped.org/), so have not moved ahead with getting these TypeScript definitions published for broader consumption.

I have updated them to the best of my knowledge to match the `meteor-integration` module, so feel free to link to them, though I would value validation that they work for for someone, else, as I have only exercised a portion of them in my project and implemented no `typings` tests.
